assignment2/2b/src/Tree.java

```diff
@@ -0,0 +1,122 @@
+    private Tree<T> right;
+
+    Tree(T value) {
+        this(value, null, null, null);
+    }
+
+    Tree(T value, Tree<T> parent, Tree<T> left, Tree<T> right) {
+        this.value = value;
+        this.parent = parent;
+        this.left = left;
+        this.right = right;
+    }
+
+    /**
+     * Add new element
+     *
+     * @param newValue  -- new element
+     * @param parentVal -- value of parent
+     * @param isLeft    -- is left or right child
+     * @return true if added
+     */
+    boolean add(T newValue, T parentVal, boolean isLeft) {
+        if (parentVal.equals(value)) {
+            if (isLeft) {
+                left = new Tree<T>(newValue, this, left, null);
+                if (left.left != null) {
+                    left.left.parent = left;
+                }
+            } else {
+                right = new Tree<T>(newValue, this, null, right);
+                if (right.right != null) {
+                    right.right.parent = right;
+                }
+            }
+            return true;
+        } else {
+            return (left != null && left.add(newValue, parentVal, isLeft))
+                    || (right != null && right.add(newValue, parentVal, isLeft));
+        }
+    }
+
+    /**
+     *
+     * Get common ancestor
+     *
+     * @param element1 -- first element
+     * @param element2 -- second element
+     * @return -- common ancestors for elements
+     * @throws TreeException.ElementDoesntFound -- if element isn't in tree
+     * @throws TreeException.NoCommonAncestor -- if one of elements is root
+     */
+    T getCommonAncestor(T element1, T element2)
+            throws TreeException.ElementDoesntFound, TreeException.NoCommonAncestor {
+        if (value.equals(element1) || value.equals(element2)) {
+            throw new TreeException.NoCommonAncestor("No common ancestor for values " + element1.toString()
+                    + " and " + element2.toString());
+        }
+
+        List<T> parents1 = getParents(element1);
+        List<T> parents2 = getParents(element2);
+        if (parents1.isEmpty() && !value.equals(element1)) {
+            throw new TreeException.ElementDoesntFound("element " + element1.toString() + " doesn't found");
+        } else if (parents2.isEmpty() && !value.equals(element2)) {
+            throw new TreeException.ElementDoesntFound("element " + element2.toString() + " doesn't found");
+        } else {
+            for (T i : parents1) {
+                if (parents2.contains(i)) {
+                    return i;
+                }
+            }
+        }
+
+        throw new TreeException.NoCommonAncestor("No common ancestor for values " + element1.toString()
+                + " and " + element2.toString());
+    }
+
+    /**
+     * Get list of parents for element. If element doesn't exist, return empty list
+     *
+     * @param element -- value of element
+     * @return -- list of ancestors
+     */
+    private List<T> getParents(T element) {
```

Comment:
As you have the explicit parent links, you can break this into two phases: Find the node, and given a node build the list of parents. This would be a little easier to read.